File: StrideApp/Models/NutritionLog.swift
import Foundation
import SwiftData

/// Local SwiftData mirror of backend nutrition_logs.
/// Source of truth is the server; we cache so Today view stays snappy offline.
@Model
final class NutritionLog {
    @Attribute(.unique) var id: String

    var loggedAt: Date
    var mealType: String          // breakfast / lunch / dinner / snack / pre_workout / post_workout / other
    var entryMethod: String       // photo / text / manual / template

    var photoUrl: String?
    var rawDescription: String?

    /// JSON-encoded `[{name, qty, calories, carbs_g, protein_g, fat_g}]`
    var itemsJSON: String

    var totalCalories: Int
    var totalCarbsG: Double
    var totalProteinG: Double
    var totalFatG: Double

    var confidence: Double?

    init(
        id: String,
        loggedAt: Date,
        mealType: String,
        entryMethod: String,
        photoUrl: String? = nil,
        rawDescription: String? = nil,
        itemsJSON: String = "[]",
        totalCalories: Int = 0,
        totalCarbsG: Double = 0,
        totalProteinG: Double = 0,
        totalFatG: Double = 0,
        confidence: Double? = nil
    ) {
        self.id = id
        self.loggedAt = loggedAt
        self.mealType = mealType
        self.entryMethod = entryMethod
        self.photoUrl = photoUrl
        self.rawDescription = rawDescription
        self.itemsJSON = itemsJSON
        self.totalCalories = totalCalories
        self.totalCarbsG = totalCarbsG
        self.totalProteinG = totalProteinG
        self.totalFatG = totalFatG
        self.confidence = confidence
    }
}

File: app/models/nutrition_log.py
import enum
import uuid
from datetime import datetime, timezone

from sqlalchemy import (
    Column, String, DateTime, Float, Integer, Boolean, Text, ForeignKey, Index, Numeric
)
from sqlalchemy.dialects.postgresql import UUID, JSONB

from app.database import Base


class MealType(str, enum.Enum):
    BREAKFAST = "breakfast"
    LUNCH = "lunch"
    DINNER = "dinner"
    SNACK = "snack"
    PRE_WORKOUT = "pre_workout"
    POST_WORKOUT = "post_workout"
    OTHER = "other"


class EntryMethod(str, enum.Enum):
    PHOTO = "photo"
    TEXT = "text"
    MANUAL = "manual"
    TEMPLATE = "template"


class NutritionLog(Base):
    """
    One row per meal/snack. Parsed items are stored as JSONB so we can show
    them line-by-line in the confirm UI and re-render later. Totals are
    denormalized for fast daily aggregation.
    """
    __tablename__ = "nutrition_logs"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True)

    logged_at = Column(DateTime(timezone=True), nullable=False, default=lambda: datetime.now(timezone.utc))
    meal_type = Column(String(20), nullable=False)             # MealType value
    entry_method = Column(String(16), nullable=False)          # EntryMethod value

    # Photo (R2 URL) and/or raw user text
    photo_url = Column(String(500), nullable=True)
    raw_description = Column(Text, nullable=True)

    # [{name, qty, calories, carbs, protein, fat, fiber}, ...]
    parsed_items = Column(JSONB, nullable=False, default=list)

    total_calories = Column(Integer, nullable=True)
    total_carbs = Column(Numeric(7, 1), nullable=True)
    total_protein = Column(Numeric(7, 1), nullable=True)
    total_fat = Column(Numeric(7, 1), nullable=True)
    total_fiber = Column(Numeric(7, 1), nullable=True)

    # Vision-parse confidence 0–1; 1.0 for manual entries
    confidence = Column(Float, nullable=True)
    edited_after_parse = Column(Boolean, nullable=False, default=False)

    # Optional link to a planned/completed Workout when this is a pre/post workout meal
    related_workout_id = Column(UUID(as_uuid=True), nullable=True)

    __table_args__ = (
        Index("ix_nutrition_user_logged", "user_id", "logged_at"),
    )
